Source/WebCore/dom/DocumentMarkerController.h

```diff
@@ -93,7 +97,15 @@ class DocumentMarkerController final : public CanMakeCheckedPtr<DocumentMarkerCo
     WeakPtr<DocumentMarker> markerContainingPoint(const LayoutPoint&, DocumentMarker::Type);
     WEBCORE_EXPORT Vector<FloatRect> renderedRectsForMarkers(DocumentMarker::Type);
 
-    WEBCORE_EXPORT void forEachOfTypes(OptionSet<DocumentMarker::Type>, const Function<bool(Node&, RenderedDocumentMarker&)>);
+    template<IterationMode = IterationMode::Forward>
+    WEBCORE_EXPORT void forEach(const SimpleRange&, OptionSet<DocumentMarker::Type>, Function<bool(Node&, RenderedDocumentMarker&)>&&);
+
+    template<>
+    WEBCORE_EXPORT void forEach<IterationMode::Forward>(const SimpleRange&, OptionSet<DocumentMarker::Type>, Function<bool(Node&, RenderedDocumentMarker&)>&&);
+
+    template<>
+    WEBCORE_EXPORT void forEach<IterationMode::Reversed>(const SimpleRange&, OptionSet<DocumentMarker::Type>, Function<bool(Node&, RenderedDocumentMarker&)>&&);
```

Comment:
Not that it can't be, but does it need to be templated? You could accept `IterationMode` as a function argument. Or if you want to avoid the runtime indirection:

```cpp
struct IterationDirection {
    struct Forward {};
    struct Backward {};
};

void forEach(const SimpleRange&, OptionSet<DocumentMarker::Type>, Function<bool(Node&, RenderedDocumentMarker&)>&&, IterationDirection::Forward);

void forEach(const SimpleRange&, OptionSet<DocumentMarker::Type>, Function<bool(Node&, RenderedDocumentMarker&)>&&, IterationDirection::Backward);
```

And then you'd change callers as such:

```cpp
markers.forEach(..., WebCore::DocumentMarkerController::IterationDirection::Forward { });

markers.forEach(..., WebCore::DocumentMarkerController::IterationDirection::Backward { });
```

Reply:
This is basically the tag dispatch pattern, by the way. https://www.fluentcpp.com/2018/04/27/tag-dispatching/

Reply:
I think I'll stick with the template approach